app/src/main/java/unipd/se18/ocrcamera/Inci.java

```diff
@@ -0,0 +1,133 @@
+ * For more informations about openCSV:
+ * http://opencsv.sourceforge.net/
+ * @author Francesco Pham
+ */
+public class Inci {
+
+    //list of recognized ingredients where are stored informations about ingredients
+    private List<Ingredient> listIngredients;
+
+    private static final String TAG = "Inci";
+
+    /*
+        threshold of similarity between the found text and the ingredient name inside db
+        below this threshold the ingredient is considered not found inside db therefore not pushed into the list
+    */
+    private static final double similarityThreshold = 0.6;
+
+    /**
+     * constructor
+     * @param inputStream inputStream of the inci database file
+     * @author Francesco Pham
+     */
+    public Inci(InputStream inputStream){
+        Reader reader = new BufferedReader(new InputStreamReader(inputStream));
+
+        listIngredients = new ArrayList<Ingredient>(); //initializing list of ingredients
+
+        //initializing openCSV reader
+        CSVReader csvReader = new CSVReader(reader);
+        String[] line;
+
+        //for each line in the csv add an Ingredient object to the list
+        try {
+            while ((line = csvReader.readNext()) != null) {
+                Ingredient element = new Ingredient();
+                element.setCosingRefNo(line[0]);
+
+                if(line.length > 1)
+                    element.setInciName(line[1]);
+                else
+                    Log.d(TAG, "inci name not found while parsing "+element.getCosingRefNo());
+
+                if(line.length > 6)
+                    element.setDescription(line[6]);
+                else
+                    Log.d(TAG, "description not found while parsing "+element.getCosingRefNo());
+
+                listIngredients.add(element);
+            }
+
+        } catch(IOException e){
+            Log.e(TAG, "Error trying to read csv");
+        }
+
+        //closing
+        try {
+            reader.close();
+            csvReader.close();
+        } catch(IOException e){
+            Log.e(TAG, "Error closing csv reader");
+        }
+    }
+
+    /**
+     * Find the best matching ingredient in the database using weighted levenshtein algorithm
+     * @param ingredient The ingredient we are looking for
+     * @return Ingredient object that contains the most similar ingredient to the text taken from ocr
+     * @author Francesco Pham
+     */
+    private Ingredient findBestMatchingIngredient(String ingredient){
+        //ignoring case by converting to upper case like all texts in database
+        ingredient = ingredient.toUpperCase();
+
+        LevenshteinStringComparator stringComparator = new LevenshteinStringComparator();
+        double maxSimilarity = -1;
+        int bestMatchingIngredient = -1;
+        for(int i = 0; i< listIngredients.size(); i++){
+            double similarity = stringComparator.getNormalizedSimilarity(listIngredients.get(i).getInciName(), ingredient);
```

Comment:
Is there a way to avoid calculating the exact similarity for each ingredient? for example, could we stop the evaluation before if we find out that we can't do better than 'similarityThreshold'? Maybe this could improve the performance

Reply:
I thinks it is better if we find the best matching ingredient and not the first ingredient that reach the threshold. We won't improve the performance much better doing that.

Reply:
I mean, if you are comparing 'aqua' with 'gas' (two words that have nothing in common) can we stop the evaluation of the levenshtein distance before (for example noticing that the distance is greater than 3) and save some time? This reasoning could be done for many other ingredients